### PR TITLE
Allow assertions to be overridden without mutating globals

### PR DESCRIPTION
**Pull request by mmun**
This is the first step for migration our assertions to be module-based.

### CODE REVIEW

**Comment by rwjblue** [comment]
Looks great!

---

This implementation requires overriding all of the methods at once, do you think there is value in allowing individual methods to be overridden?

**Comment by mmun** [comment]
I can do an override for each one instead.

**Comment by rwjblue** [comment]
Totally up to you, I'm happy to merge as is.

I was just curious what you thought since most of the use cases we have right now involve only overriding `Ember.deprecate` and `Ember.assert`.

**Comment by rwjblue** [comment]
Looks like this needs a rebase...

File: packages/ember-metal/lib/assert.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+export let debugFunctions = {
+  assert() {},
+  warn() {},
+  debug() {},
+  deprecate() {},
+  deprecateFunc(_, fn) { return fn; },
```
Comment (rwjblue) [inline packages/ember-metal/lib/assert.js]: We just fixed the default (prod) version of this this morning, can you update to ensure the last arg is always returned?

**Pull request merged**